pkg/api/v1/clients/wrapper/multi_watcher.go

```diff
@@ -0,0 +1,182 @@
+func NewWatchAggregator() WatchAggregator {
+	sources := make(map[clients.ResourceWatcher][]removeWatch)
+	sinks := make(map[resourceSink]addWatch)
+	watchers := make(map[clients.ResourceWatcher]struct{})
+	return &watchAggregator{sources: sources, sinks: sinks, watchers: watchers}
+}
+
+func (c *watchAggregator) Watch(namespace string, opts clients.WatchOpts) (<-chan resources.ResourceList, <-chan error, error) {
+	opts = opts.WithDefaults()
+	// create new sinks for this watch
+	out := make(chan resources.ResourceList)
+	aggregatedErrs := make(chan error)
+
+	// a shared map that will be used to merge resources from different watchers
+	listsByWatcher := make(resourcesByWatcher)
+	access := sync.RWMutex{}
+
+	// construct a func for adding an input watcher to this sink
+	addWatch := func(watcher clients.ResourceWatcher) (err error) {
+		// this function starts a watch for the watcher using the root context
+		// we want to cancel it if :
+		// the root context is cancelled
+		// the watcher is removed
+		ctx, cancel := context.WithCancel(opts.Ctx)
+
+		// start a watch for the watcher on this namespace
+		source, errs, err := watcher.Watch(namespace, clients.WatchOpts{Ctx: ctx, RefreshRate: opts.RefreshRate})
+		if err != nil {
+			return err
+		}
+
+		// read lists from the source channel,
+		// group its resources by type
+		go func() {
+			defer cancel()
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case err := <-errs:
+					// if the source starts returning errors, remove its list from the snasphot
+					access.Lock()
+					delete(listsByWatcher, watcher)
+					access.Unlock()
+					aggregatedErrs <- err
+					select {
+					case <-ctx.Done():
+						return
+					case out <- listsByWatcher.merge():
+					}
+				case list, ok := <-source:
+					if !ok {
+						return
+					}
+					// add/update the list to the snapshot
+					access.Lock()
+					listsByWatcher[watcher] = list
+					access.Unlock()
+					select {
+					case <-ctx.Done():
+						return
+					case out <- listsByWatcher.merge():
+					}
+				}
+			}
+		}()
+
+		// construct a function for removing this watcher from this sink
+		removeWatch := func(watcher clients.ResourceWatcher) {
+			// remove the watcher+resources from the snapshot
+			access.Lock()
+			delete(listsByWatcher, watcher)
+			access.Unlock()
+			cancel()
+		}
+
+		c.sourcesAccess.Lock()
+		c.sources[watcher] = append(c.sources[watcher], removeWatch)
+		c.sourcesAccess.Unlock()
+
+		return nil
+	}
+
+	// add all the registered watchers to the sink
+	c.watchersAccess.RLock()
+	for w := range c.watchers {
+		if err := addWatch(w); err != nil {
+			return nil, nil, err
+		}
+	}
+	c.watchersAccess.RUnlock()
+
+	// store a way to add watches to this sink
+	c.sinksAccess.Lock()
+	c.sinks[out] = addWatch
+	c.sinksAccess.Unlock()
+
+	go func() {
+		// context is closed, clean up watch resources
+		<-opts.Ctx.Done()
+		c.sinksAccess.Lock()
+		delete(c.sinks, out)
+		c.sinksAccess.Unlock()
+		close(out)
```

Comment:
this might panic if another goroutine is just about to send something on the out channel. (before it detects the context was cancelled)

Reply:
you can use a waitgroup to guarantee that no other go routine is using the channel